File: String/065.Valid_Number/65.Valid_Number.py
class Solution:
    def isNumber(self, s: str) -> bool:
        state = [{},
                {"blank": 1, "sign": 2, "digit": 3, "dot": 4}, # s1
                {"digit": 3, "dot": 4}, # s2
                {"digit": 3, "dot": 5, "e": 6, "blank": 9}, # s3
                {"digit": 5}, # s4
                {"digit": 5, "e": 6, "blank": 9}, # s5
                {"sign": 7, "digit": 8}, # s6
                {"digit": 8}, # s7
                {"digit": 8, "blank": 9}, # s8
                {"blank": 9}] # s9
        
        cur_state = 1
        for c in s :
            if c == ' ' :
                c = "blank"
            elif c == '+' or c == '-' :
                c = "sign"
            elif c == '.' :
                c = "dot"
            elif c == 'e' :
                c = "e"
            elif c >= '0' and c <= '9' :
                c = "digit"
            
            if c not in state[cur_state].keys() :
                return False
            cur_state = state[cur_state][c]
            
        if  cur_state not in [3, 5, 8, 9] : # end state
            return False
        return True

File: String/065.Valid_Number/65.Valid_Number.cpp
class Solution {
public:
    bool isNumber(string s) {
        if (s.size() == 0) return false;
        string new_s = removespace(s);
        removeuary(new_s);
        if (new_s.size() == 0) return false;
        
        vector<string> parts(2);
        if (new_s.find('e') != string::npos) {
            int f = new_s.find_first_of('e');
            int e = new_s.find_last_of('e');
            if (f != e) return false;
            parts[0] = new_s.substr(0, f);
            parts[1] = new_s.substr(f + 1, new_s.size() - 1 - f);
            removeuary(parts[1]);
            if (parts[0].size() == 0 || parts[1].size() == 0) return false;
        }
        else {
            parts[0] = new_s;
        }
        
        for (int k = 0; k < 2; k++) {
            int dot = 0;
            for (int i = 0; i < parts[k].size(); i++) {
                if (isdigit(parts[k][i])) continue;
                else if (parts[k][i] == '.') {
                    if (k == 0) {
                        if (dot == 1 || parts[k].size() == 1) return false;
                        else {
                            dot = 1;
                            if (i == 0 || i == parts[k].size() - 1) continue;
                            if (!isdigit(parts[k][i - 1]) || !isdigit(parts[k][i + 1])) return false;
                        }
                    }
                    else return false;
                }
                else return false;
            }
        }
        return true;
    }
    
    string removespace(string s) {
        int i = 0, j = s.size() - 1;
        while (s[i] == ' ' && i < j) i++;
        while (s[j] == ' ' && i <= j) j--;
        return s.substr(i, j - i + 1);
    }
    
    void removeuary(string &s) {
        if (s.size() == 0) return;
        if (s[0] == '+' || s[0] == '-') {
            s.erase(0, 1);
        }
    }
};